wagtail/admin/edit_handlers.py

```diff
@@ -789,8 +792,89 @@ def render(self):
         )
 
 
+class CommentPanel(EditHandler):
+    def __init__(self, *args, **kwargs):
+        self.comments = []
+        super().__init__(*args, **kwargs)
+
+    def required_formsets(self):
+        # add the comments formset
+        # we need to pass in the current user for validation on the formset
+        # this could alternatively be done on the page form itself if we added the
+        # comments formset there, but we typically only add fields via edit handlers
+        current_user = getattr(self.request, 'user', None)
+
+        class CommentReplyFormWithRequest(CommentReplyForm):
+            user = current_user
+
+        class CommentFormWithRequest(CommentForm):
+            user = current_user
+
+            class Meta:
+                formsets = {
+                    'replies': {
+                        'form': CommentReplyFormWithRequest
+                    }
+                }
+
+        return {
+            'comments': {
+                'form': CommentFormWithRequest,
+                'fields': ['text', 'contentpath'],
+            }
+        }
+
+    template = "wagtailadmin/edit_handlers/comments/comment_panel.html"
+    js_template = "wagtailadmin/edit_handlers/comments/comment_panel.js"
+    declarations_template = "wagtailadmin/edit_handlers/comments/comment_declarations.html"
+
+    def html_declarations(self):
+        return render_to_string(self.declarations_template)
+
+    def on_form_bound(self):
+        self.comments = []
+        comment_forms = self.form.formsets['comments'].forms
+        for form in comment_forms:
+            instance = form.instance
+            instance.replies = [reply_form.instance for reply_form in form.formsets['replies'].forms]
+            self.comments.append(instance)
+
+    def get_context(self):
+        user = getattr(self.request, 'user', None)
+        authors = {user.pk: user_display_name(user)} if user else {}
+        serialized_comments = []
+        for comment in self.comments:
```

Comment:
**Thought:** If I could only do something like this... I don't quite get whether all of this is coming from a database or not.

```python
pks = [comment.pk for comment in self.comments]
qs = Comment.objects.filter(pk__in=pks).select_related('user').prefetch_related('replies__user').in_bulk()
for qs[pk] in pks:
```
Alternatively maybe prefetch the users in one go to save some queries.

```python
user_pks = []
for comment in self.comments:
    user_pks.append(comment.user_id)
     # all() should hopefully mean this will be cached for the future if you don't use filter etc.
    user_pks += [reply.user_id for reply in comment.replies.all()]
users = get_user_model().objects.filter(pk__in=user_pks).in_bulk()

# Down the line
users[reply.user_id]
users[comment.user_id]
```

Reply:
The first one I think won't work because `self.comments` can include unsaved edited comments from the form (not the db), so fetching the comments from the db isn't guaranteed to be the same, but you're right that I could optimise more here! I'll have an experiment with this (I did originally have some prefetching, I'll see if I can bring that back in)

Reply:
I've added your second suggestion. Unfortunately `prefetch_related_objects` didn't seem to work for this.